File: engines/ts/src/predicates.ts
import { deepEq } from "./canonical.js";
import type { Predicate, Snapshot } from "./types.js";

/** Internal sentinel — distinct from JSON null. Never serialized. */
export const MISSING: unique symbol = Symbol("fausth.MISSING");

export type PathValue = unknown | typeof MISSING;

export function getPath(snapshot: Snapshot, path: string): PathValue {
  const parts = path.split(".");
  let cur: unknown = snapshot;
  for (const p of parts) {
    if (cur === null || cur === undefined || typeof cur !== "object") {
      return MISSING;
    }
    const obj = cur as Record<string, unknown>;
    if (!(p in obj)) {
      return MISSING;
    }
    cur = obj[p];
  }
  return cur;
}

export function isMissing(v: PathValue): v is typeof MISSING {
  return v === MISSING;
}

export function evalPredicate(pred: Predicate, snapshot: Snapshot): boolean {
  if ("all" in pred) {
    return pred.all.every((p) => evalPredicate(p, snapshot));
  }
  if ("any" in pred) {
    return pred.any.some((p) => evalPredicate(p, snapshot));
  }
  if ("not" in pred) {
    return !evalPredicate(pred.not, snapshot);
  }
  const value = getPath(snapshot, pred.path);
  if ("eq_path" in pred) {
    const other = getPath(snapshot, pred.eq_path);
    if (isMissing(value) && isMissing(other)) return true;
    if (isMissing(value) || isMissing(other)) return false;
    return deepEq(value, other);
  }
  if ("eq" in pred) {
    if (isMissing(value)) return false;
    return deepEq(value, pred.eq);
  }
  if ("neq" in pred) {
    // Missing is absent, not "not equal"
    if (isMissing(value)) return false;
    return !deepEq(value, pred.neq);
  }
  if (isMissing(value) || typeof value !== "number" || !Number.isInteger(value)) {
    return false;
  }
  if ("lt" in pred) return value < pred.lt;
  if ("lte" in pred) return value <= pred.lte;
  if ("gt" in pred) return value > pred.gt;
  if ("gte" in pred) return value >= pred.gte;
  return false;
}

File: engines/py/fausth/predicates.py
from __future__ import annotations

from typing import Any

from .canonical import deep_eq

MISSING = object()


def get_path(snapshot: dict[str, Any], path: str) -> Any:
    cur: Any = snapshot
    for p in path.split("."):
        if not isinstance(cur, dict) or p not in cur:
            return MISSING
        cur = cur[p]
    return cur


def is_missing(v: Any) -> bool:
    return v is MISSING


def eval_predicate(pred: dict[str, Any], snapshot: dict[str, Any]) -> bool:
    if "all" in pred:
        return all(eval_predicate(p, snapshot) for p in pred["all"])
    if "any" in pred:
        return any(eval_predicate(p, snapshot) for p in pred["any"])
    if "not" in pred:
        return not eval_predicate(pred["not"], snapshot)
    path = pred["path"]
    value = get_path(snapshot, path)
    if "eq_path" in pred:
        other = get_path(snapshot, pred["eq_path"])
        if is_missing(value) and is_missing(other):
            return True
        if is_missing(value) or is_missing(other):
            return False
        return deep_eq(value, other)
    if "eq" in pred:
        if is_missing(value):
            return False
        return deep_eq(value, pred["eq"])
    if "neq" in pred:
        if is_missing(value):
            return False
        return not deep_eq(value, pred["neq"])
    if is_missing(value) or not isinstance(value, int) or isinstance(value, bool):
        return False
    if "lt" in pred:
        return value < pred["lt"]
    if "lte" in pred:
        return value <= pred["lte"]
    if "gt" in pred:
        return value > pred["gt"]
    if "gte" in pred:
        return value >= pred["gte"]
    return False
